### DISCUSSION TITLE
Burning our stablecoins before redeeming in fuzz testing

### DISCUSSION THREAD

**Discussion by Podziuu** [Cyfrin/foundry-full-course-cu#2117] [Ideas]
Hello,

While writing fuzz tests for the Stablecoin project, I discovered an edge case where the `redeemCollateral` function reverts due to the user’s health factor being broken. Upon investigation, I found that the user had minted some coins, and the `bound` function selected almost the maximum amount to redeem collateral. This caused the revert, which is expected behavior.

To handle this, I modified the `redeemCollateral` function to pass all cases. Here is the updated function:
``` solidity
function redeemCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
        ERC20Mock collateral = _getCollateralFromSeed(collateralSeed);
        uint256 maxAmountToRedeem = engine.getCollateralBalance(msg.sender, address(collateral));
        (uint256 mintedPdn,) = engine.getAccountInformation(msg.sender);

        amountCollateral = bound(amountCollateral, 0, maxAmountToRedeem);

        if (amountCollateral == 0) {
            return;
        }
        vm.startPrank(msg.sender);
        if (mintedPdn > 0) { // burn all minted coins to secure redeeming
            pdn.approve(address(engine), mintedPdn);
            engine.burnPdn(mintedPdn);
        }
        engine.redeemCollateral(address(collateral), amountCollateral);
        vm.stopPrank();
    }
```
Is this a good approach, or am I missing something?

**Comment by EngrPips** [comment]
Hey @Podziuu, This seems like a good approach. We usually want our invariant test suite to break our protocol in a way that we don't foresee, so we should write the invariant test suite to prevent all reverts we know of.

**Reply by Podziuu** [reply]
Thanks for the confirmation @EngrPips

**Reply by EngrPips** [reply]
You are welcome, Buddy. Keep crushing it.
